spec/importeur/data_sources/rocketfuel_spec.rb

```diff
@@ -9,12 +9,13 @@
   let(:resource_resource)  { instance_double(RocketfuelApi::Resource::Company) }
 
   before do
-    expect(rocketfuel_service).to receive(:get_all).and_return([resource_resource])
+    allow(rocketfuel_service).to receive(:get_all).and_return([resource_resource])
   end
 
   describe '#items' do
     it 'returns items' do
       expect(data_source.items).to eq([resource_resource])
+      expect(rocketfuel_service).to have_received(:get_all).with({})
```

Comment:
By the name `get_all` I assume this method doesn't have side-effects, so we don't need to "expect" it. What we care about is that `#items` returns correct value depending on the params. And we can test it:
```ruby
before do
  allow(rocketfuel_service).to receive(:get_all).with({}).and_return([resource_resource])
  allow(rocketfuel_service).to receive(:get_all).with(params).and_return(specific_items)
end

it 'returns items for default params' do
  expect(data_source.items).to eq([resource_resource])
end

context 'with params' do
  …
  it 'returns items for custom params' do
   expect(data_source.items).to eq(specific_items)
  end
end
```

Reply:
The last row, for outgoing messages, we only need to "expect" command messages, which change something, but not queries, which just return data.
![](http://jnoconor.github.io/images/unit-testing-chart-sandi-metz.png)

Reply:
Interesting. I had to look for the presentation to understand it though :wink:

Reply:
Yeah, I suspected the image itself might not be too descriptive.

Reply:
There's also a book from the author, http://www.poodr.com/. One of those actually good books :) And quite short, also.